keras/layers/attention/group_query_attention.py

```diff
@@ -0,0 +1,219 @@
+    tensor.
+
+    Args:
+        head_dim: Size of each attention head.
+        num_query_heads: Number of query attention heads.
+        num_key_value_heads: Number of key and value attention heads.
+        use_bias: Boolean, whether the dense layers use bias vectors/matrices.
+
+    Call arguments:
+        query: Query tensor of shape `(batch_dim, target_seq_len, feature_dim)`,
+            where `batch_dim` is batch size, `target_seq_len` is the length of
+            target sequence, and `feature_dim` is dimension of feature.
+        value: Value tensor of shape `(batch_dim, source_seq_len, feature_dim)`,
+            where `batch_dim` is batch size, `source_seq_len` is the length of
+            source sequence, and `feature_dim` is dimension of feature.
+        key: Optional key tensor of shape
+            `(batch_dim, source_seq_len, feature_dim)`. If not given, will use
+            `value` for both `key` and `value`, which is most common case.
+        mask: A boolean mask of shape
+            `(batch_dim, target_seq_len, source_seq_len)`, that prevents
+            attention to certain positions. The boolean mask specifies which
+            query elements can attend to which key elements, where 1 indicates
+            attention and 0 indicates no attention. Broadcasting can happen for
+            the missing batch dimensions and the head dimension.
+        return_attention_scores: A boolean to indicate whether the output
+            should be `(attention_output, attention_scores)` if `True`, or
+            `attention_output` if `False`. Defaults to `False`.
+        training: Python boolean indicating whether the layer should behave in
+            training mode (adding dropout) or in inference mode (no dropout).
+            Will go with either using the training mode of the parent
+            layer/model or `False` (inference) if there is no parent layer.
+
+    Returns:
+        attention_output: Result of the computation, of shape
+            `(batch_dim, target_seq_len, feature_dim)`, where `target_seq_len`
+            is for target sequence length and `feature_dim` is the query input
+            last dim.
+        attention_scores: (Optional) attention coefficients.
+    """
+
+    def __init__(
+        self,
+        head_dim,
+        num_query_heads,
+        num_key_value_heads,
+        use_bias=False,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.head_dim = head_dim
+        self.num_query_heads = num_query_heads
+        self.num_key_value_heads = num_key_value_heads
+        self.num_repeats = num_query_heads // num_key_value_heads
+        self.use_bias = use_bias
+
+    def build(
+        self,
+        query_shape,
+        value_shape,
+        key_shape=None,
+    ):
+        key_shape = value_shape if key_shape is None else key_shape
+        self.feature_dim = query_shape[-1]
+        self._query_dense = Dense(
```

Comment:
If it's useful. Here's a draft on a multi-query attention build block I was using for a model port. This would need adapting for grouped query (key and value should still output 4d).

```python
        # Einsum variables:
        # b = batch size
        # q = query length
        # k = key/value length
        # m = model dim
        # n = num heads
        # h = head dim
        model_dim = query_shape[-1]
        head_dim = model_dim // self.num_heads
        key_shape = value_shape if key_shape is None else key_shape
        self.query_dense = keras.layers.EinsumDense(
            "bqm,mnh->bqnh",
            output_shape=(None, self.num_heads, head_dim),
            kernel_initializer="variance_scaling",
            name="query_dense",
        )
        self.query_dense.build(query_shape)
        self.value_dense = keras.layers.EinsumDense(
            "bkm,mh->bkh",
            output_shape=(None, head_dim),
            kernel_initializer="variance_scaling",
            name="value_dense",
        )
        self.value_dense.build(value_shape)
        self.key_dense = keras.layers.EinsumDense(
            "bkm,mh->bkh",
            output_shape=(None, head_dim),
            kernel_initializer="variance_scaling",
            name="key_dense",
        )
        self.key_dense.build(key_shape)
        self._dot_product_equation = "bqnh,bkh->bnqk"
        self._combine_equation = "bnqk,bkh->bqnh"
        self.output_dense = keras.layers.EinsumDense(
            "bqnh,nhm->bqm",
            output_shape=(None, model_dim),
            name="output_dense",
        )
        self.output_dense.build((None, None, self.num_heads, head_dim))
        self.softmax_layer = keras.layers.Softmax()
        self.built = True
```

Reply:
Thank you. I apologize for the delayed response. Unfortunately, I was caught up in a conference deadline.

Reply:
All good! No rush. Just realized I had some code laying around that might be useful here.

Reply:
@mattdangerw Apologies for the delay once again; I've just returned home after attending the conference abroad.

I have updated the code to incorporate all your feedback. Thanks for the code snippet; they were helpful.

Since this is my first time working with `einsum`, it would be helpful if you could double-check the equations and also confirm if the `bias_axes` are correct.

Next, I will add the unit tests. Additionally, I couldn't find the instructions for `bazel` tests that used to be in this repository. Could you please point me to that page?

Reply:
@mattdangerw  unit tests have also been added.